File: builder-pattern/go/src/Director.go
package src

// 指挥调度类，负责利用建造者建造产品，隔离需求
type Director struct {
}

// 建造phone1
func (c *Director) BuildIPhone(builder Builder) {
  builder.Reset()
  var screen = []int{120, 500}
  builder.SetName("iPhone")
  builder.SetScreen(screen)
  builder.SetGPU(100)
}

// 建造phone2
func (c *Director) BuildHuaweiPhone(builder Builder) {
  builder.Reset()
  var screen = []int{140, 600}
  builder.SetName("HuaweiPhone")
  builder.SetScreen(screen)
  builder.SetGPU(102)
}

// 建造phone3
func (c *Director) BuildMiPhone(builder Builder) {
  builder.Reset()
  var screen = []int{130, 550}
  builder.SetName("MiPhone")
  builder.SetScreen(screen)
  builder.SetGPU(103)
}

File: builder-pattern/java/src/Director.java
package src;

// 指挥调度类，负责利用建造者建造产品，隔离需求
public class Director {

  // 建造phone1
  public void buildIPhone(Builder builder) {
    builder.reset();
    Integer[] screen = { 120, 500 };
    builder.setName("iPhone");
    builder.setScreen(screen);
    builder.setGPU(100);
  }

  // 建造phone2
  public void buildHuaweiPhone(Builder builder) {
    builder.reset();
    Integer[] screen = { 130, 600 };
    // 也可以链式调用
    builder.setName("HuaweiPhone")
    .setScreen(screen)
    .setGPU(102);
  }

  // 建造phone3
  public void buildMiPhone(Builder builder) {
    builder.reset();
    Integer[] screen = { 120, 650 };
    builder.setName("MiPhone");
    builder.setScreen(screen);
    builder.setGPU(103);
  }

}
